browser/modules/PlacesProvider.jsm

```diff
@@ -0,0 +1,258 @@
+  "resource://gre/modules/PlacesUtils.jsm");
+
+XPCOMUtils.defineLazyGetter(this, "EventEmitter", function() {
+  const {EventEmitter} = Cu.import("resource://gre/modules/devtools/event-emitter.js", {});
+  return EventEmitter;
+});
+
+XPCOMUtils.defineLazyGetter(this, "gPrincipal", function() {
+  let uri = Services.io.newURI("about:newtab", null, null);
+  return Services.scriptSecurityManager.getNoAppCodebasePrincipal(uri);
+});
+
+// The maximum number of results PlacesProvider retrieves from history.
+const HISTORY_RESULTS_LIMIT = 100;
+
+/**
+ * Singleton that checks if a given link should be displayed on about:newtab
+ * or if we should rather not do it for security reasons. URIs that inherit
+ * their caller's principal will be filtered.
+ */
+let LinkChecker = {
+  _cache: {},
+
+  get flags() {
+    return Ci.nsIScriptSecurityManager.DISALLOW_INHERIT_PRINCIPAL |
+           Ci.nsIScriptSecurityManager.DONT_REPORT_ERRORS;
+  },
+
+  checkLoadURI: function LinkChecker_checkLoadURI(aURI) {
+    if (!(aURI in this._cache)) {
+      this._cache[aURI] = this._doCheckLoadURI(aURI);
+    }
+
+    return this._cache[aURI];
+  },
+
+  _doCheckLoadURI: function LinkChecker_doCheckLoadURI(aURI) {
+    try {
+      Services.scriptSecurityManager.
+        checkLoadURIStrWithPrincipal(gPrincipal, aURI, this.flags);
+      return true;
+    } catch (e) {
+      // We got a weird URI or one that would inherit the caller's principal.
+      return false;
+    }
+  }
+};
+
+/**
+ * Singleton that provides utility functions for links.
+ * A link is a plain object that looks like this:
+ *
+ * {
+ *   url: "http://www.mozilla.org/",
+ *   title: "Mozilla",
+ *   frecency: 1337,
+ *   lastVisitDate: 1394678824766431,
+ * }
+ */
+let LinkUtils = {
+  _sortProperties: [
+    "frecency",
+    "lastVisitDate",
+    "url",
+  ],
+
+  /**
+   * Compares two links.
+   *
+   * @param {String} aLink1 The first link.
+   * @param {String} aLink2 The second link.
+   * @return {Number} A negative number if aLink1 is ordered before aLink2, zero if
+   *         aLink1 and aLink2 have the same ordering, or a positive number if
+   *         aLink1 is ordered after aLink2.
+   *         Order is ascending.
+   */
+  compareLinks: function LinkUtils_compareLinks(aLink1, aLink2) {
+    for (let prop of this._sortProperties) {
+      if (!(prop in aLink1) || !(prop in aLink2)) {
+        throw new Error("Comparable link missing required property: " + prop);
+      }
+    }
+    return aLink2.frecency - aLink1.frecency ||
+           aLink2.lastVisitDate - aLink1.lastVisitDate ||
+           aLink1.url.localeCompare(aLink2.url);
+  },
+};
+
+/**
+ * Singleton that serves as the default link provider for the grid. It queries
+ * the history to retrieve the most frequently visited sites.
+ */
+let Links = {
+  /**
+   * EventEmitter interface
+   */
+  eventEmitter: new EventEmitter(),
+  on: (...params) => this.eventEmitter.on(...params),
```

Comment:
This is cool, but won't work as expected because of the way that fat arrow binding works. If you check `this`, it's actually the backstage pass, not Links. What you actually want here is to inherit from EventEmmiter. You also see the same problem below, where you do `Links.eventEmitter.emit`. 

 So:
```JS
let Links = Object.create(new EventEmmiter(), {
   ...props...
}); 
```

Really sorry, but going to have to R- here :cry::cry::cry:.